raiden_libs/blockchain.py

```diff
@@ -242,41 +250,48 @@ def filter_events(self, filter_params: dict, name_to_callback: dict) -> None:
+            (self.unconfirmed_head_number, current_block)
+        )
+        # here we should probably have a callback or a user-overriden method
+        self.unconfirmed_head_number = self.confirmed_head_number
+        self.unconfirmed_head_hash = self.confirmed_head_hash
+
+    def reset_unconfirmed_on_reorg(self, current_block: int):
         """Test if chain reorganization happened (head number used in previous pass is greater than
         current_block parameter) and in that case reset unconfirmed event list."""
         if self.wait_sync_event.is_set():  # but not on first sync
-            if current_block < self.unconfirmed_head_number:
-                log.info('Chain reorganization detected. '
-                         'Resyncing unconfirmed events (unconfirmed_head=%d) [@%d]' %
-                         (self.unconfirmed_head_number, self.web3.eth.blockNumber))
-                # here we should probably have a callback or a user-overriden method
-                self.unconfirmed_head_number = self.confirmed_head_number
-                self.unconfirmed_head_hash = self.confirmed_head_hash
-            try:
-                # raises if hash doesn't exist (i.e. block has been replaced)
-                self.web3.eth.getBlock(self.unconfirmed_head_hash)
-            except ValueError:
-                log.info(
-                    'Chain reorganization detected. '
-                    'Resyncing unconfirmed events (unconfirmed_head=%d) [@%d]. '
-                    '(getBlock() raised ValueError)' %
-                    (self.unconfirmed_head_number, current_block)
-                )
-                self.unconfirmed_head_number = self.confirmed_head_number
-                self.unconfirmed_head_hash = self.confirmed_head_hash
 
-            # in case of reorg longer than confirmation number fail
+            # block number increased or stayed the same
+            if current_block >= self.unconfirmed_head_number:
+                # if the hash of our head changed, there was a chain reorg
+                current_unconfirmed_hash = self.web3.eth.getBlock(
+                    self.unconfirmed_head_number
+                ).hash
+                if current_unconfirmed_hash != self.unconfirmed_head_hash:
+                    self._detected_chain_reorg(current_block)
+            # block number decreased, there was a chain reorg
+            elif current_block < self.unconfirmed_head_number:
+                self._detected_chain_reorg(current_block)
+
+            # now we have to check that the confirmed_head_hash stayed the same
+            # otherwise the program aborts
             try:
-                self.web3.eth.getBlock(self.confirmed_head_hash)
-            except ValueError:
+                current_head_hash = self.web3.eth.getBlock(self.confirmed_head_number).hash
+                if current_head_hash != self.confirmed_head_hash:
+                    log.critical('Events considered confirmed have been reorganized')
```

Comment:
Could you add more info to the log, block numbers and hashes in particular? It'll make it easier to find out what went wrong if this scenario happens.